csp/impl/wiring/edge.py

```diff
@@ -120,6 +128,149 @@ def __invert__(self):
 
+    def __neg__(self):
+        import csp
+
+        return csp.neg(self)
+
+    # def __ceil__(self):
+    # def __floor__(self):
+    # def __round__(self):
+    # def __trunc__(self):
+    # def __lshift__(self):
+    # def __rshift__(self):
+    # def __pos__(self):
+    # def __xor__(self):
+
+    def __abs__(self):
+        import csp
+
+        return csp.abs(self)
+
+    def abs(self):
+        import csp
+
+        return csp.abs(self)
+
+    def ln(self):
+        import csp
+
+        return csp.ln(self)
+
+    def log2(self):
+        import csp
+
+        return csp.log2(self)
+
+    def log10(self):
+        import csp
+
+        return csp.log10(self)
+
+    def exp(self):
+        import csp
+
+        return csp.exp(self)
+
+    def sin(self):
+        import csp
+
+        return csp.sin(self)
+
+    def cos(self):
+        import csp
+
+        return csp.cos(self)
+
+    def tan(self):
+        import csp
+
+        return csp.tan(self)
+
+    def arcsin(self):
+        import csp
+
+        return csp.arcsin(self)
+
+    def arccos(self):
+        import csp
+
+        return csp.arccos(self)
+
+    def arctan(self):
+        import csp
+
+        return csp.arctan(self)
+
+    def sqrt(self):
+        import csp
+
+        return csp.sqrt(self)
+
+    def erf(self):
+        import csp
+
+        return csp.erf(self)
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        if ufunc == np.add:
+            if isinstance(inputs[0], Edge):
+                return inputs[0].__add__(inputs[1])
+            else:
+                return inputs[1].__add__(inputs[0])
+        elif ufunc == np.subtract:
+            if isinstance(inputs[0], Edge):
+                return inputs[0].__sub__(inputs[1])
+            else:
+                return inputs[1].__sub__(inputs[0])
+        elif ufunc == np.multiply:
+            if isinstance(inputs[0], Edge):
+                return inputs[0].__mul__(inputs[1])
+            else:
+                return inputs[1].__mul__(inputs[0])
+        elif ufunc == np.divide:
+            if isinstance(inputs[0], Edge):
+                return inputs[0].__truediv__(inputs[1])
+            else:
+                return inputs[1].__truediv__(inputs[0])
+        elif ufunc == np.floor_divide:
+            if isinstance(inputs[0], Edge):
+                return inputs[0].__floordiv__(inputs[1])
+            else:
+                return inputs[1].__floordiv__(inputs[0])
+        elif ufunc == np.power:
+            return inputs[0].pow(inputs[1])
+        elif ufunc == np.abs:
+            return inputs[0].abs()
+        elif ufunc == np.log:
+            return inputs[0].ln()
+        elif ufunc == np.log2:
+            return inputs[0].log2()
+        elif ufunc == np.log10:
+            return inputs[0].log10()
+        elif ufunc == np.exp:
+            return inputs[0].exp()
+        elif ufunc == np.exp2:
```

Comment:
Instead of this huge if/elif block, just have a dict which maps the ufunc to the operation. So ufunc_map[np.exp2] = csp.exp2. Then don't call them as member functions, call them as nodes (like you do in the member function itself). Thus, you can just `return ufunc_map[ufunc]( inputs[0] )` and avoid this block (same comment applies for the binary functions above).

Reply:
done, can't use match/case yet 😢